adi/adar1000.py

```diff
@@ -125,60 +128,78 @@ def __init__(self, uri="", beams="BEAM0"):
-        )
+        """rx_hardwaregains: Set all gains applied to RX path"""
+        self._set_attribute_dictionary("hardwaregain", False, values)
 
     @property
     def tx_phases(self):
-        """tx_phases: Set all phases of TX path"""
-        return [
-            self._get_iio_attr(b, "phase", True, dev)
-            for dev in self._ctrls
-            for b in self._beam_channels
-        ]
+        """tx_phases: Get all phases of TX path"""
+        return self._get_attribute_dictionary("phase", True)
 
     @tx_phases.setter
     def tx_phases(self, values):
-        self._set_iio_attr_float_multi_dev(
-            self._beam_channels, "phase", True, values, self._ctrls
-        )
+        """tx_phases: Set all phases of TX path"""
+        self._set_attribute_dictionary("phase", True, values)
 
     @property
     def rx_phases(self):
-        """rx_phases: Set all phases of RX path"""
-        return [
-            self._get_iio_attr(b, "phase", False, dev)
-            for dev in self._ctrls
-            for b in self._beam_channels
-        ]
+        """rx_phases: Get all phases of RX path"""
+        return self._get_attribute_dictionary("phase", False)
 
     @rx_phases.setter
     def rx_phases(self, values):
-        self._set_iio_attr_float_multi_dev(
-            self._beam_channels, "phase", False, values, self._ctrls
-        )
+        """rx_phases: Set all phases of RX path"""
+        self._set_attribute_dictionary("phase", False, values)
+
+    def _get_attribute_dictionary(self, attribute_name, output):
+        """ Get a dictionary of all attached requested attributes by name and output style """
+        return {
+            f'{dev.attrs["label"].value}_{b}': self._get_iio_attr(
+                b, attribute_name, output, dev
+            )
+            for dev in self._ctrls
+            for b in self._beam_channels
+        }
+
+    def _set_attribute_dictionary(self, attribute_name, output, values):
+        """ Set all requested attributes by name and output style using a dictionary """
+
+        # If all the new values are the same as the first value, set them all at once
+        if all([next(iter(values.values())) == v for v in values.values()]):
+            self._set_iio_attr_float_multi_dev(
+                self._beam_channels,
+                attribute_name,
+                output,
+                list(values.values()),
+                self._ctrls,
+            )
+
+        # If there are different values, set them one at a time.
+        else:
+            for key, value in values.items():
+                # Pull out the beam and channel names
+                label_sections = key.lower().split("_")
+                beam = "_".join(label_sections[:2])
+                channel = label_sections[-1]
+
+                # Determine the device
+                dev = [d.attr["label"].value.lower() == beam for d in self._ctrls][0]
```

Comment:
1. d.attr -> d.attrs
2. This returns a boolean. It needs to return a iio_device object
3. Needs a check to make sure dev is not null (otherwise set_iio_attr will default to self._ctrl)

Reply:
Yep, good catch. I'll fix this now.